File: platforms/android/runtime/core/src/main/java/io/nativeblocks/runtime/lib/NativeSqliteStore.kt
package io.nativeblocks.runtime.lib

import android.content.ContentValues
import android.content.Context
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteOpenHelper
import io.nativeblocks.runtime.engine.CacheProvider
import io.nativeblocks.runtime.engine.ErrorType

internal class NativeSqliteHelper(
    context: Context,
    databaseName: String,
) : SQLiteOpenHelper(context.applicationContext, databaseName, null, DATABASE_VERSION) {

    override fun onCreate(db: SQLiteDatabase) {
        db.execSQL(
            "CREATE TABLE IF NOT EXISTS $TABLE (" +
                    "$COLUMN_KEY TEXT PRIMARY KEY, " +
                    "$COLUMN_VALUE BLOB NOT NULL, " +
                    "$COLUMN_EXPIRY INTEGER" +
                    ")"
        )
    }

    override fun onUpgrade(db: SQLiteDatabase, oldVersion: Int, newVersion: Int) {
        db.execSQL("DROP TABLE IF EXISTS $TABLE")
        onCreate(db)
    }

    companion object {
        private const val DATABASE_VERSION = 2
        const val TABLE = "nativeblocks_cache"
        const val COLUMN_KEY = "cache_key"
        const val COLUMN_VALUE = "cache_value"
        const val COLUMN_EXPIRY = "cache_expiry"
    }
}

internal class SqliteCacheProvider(
    private val helper: NativeSqliteHelper,
) : CacheProvider {

    override fun saveBytes(key: String, value: ByteArray, ttlMillis: Long?) = uniffiCallbackGuard(ErrorType.CACHE) {
        val values = ContentValues().apply {
            put(NativeSqliteHelper.COLUMN_KEY, key)
            put(NativeSqliteHelper.COLUMN_VALUE, value)
            if (ttlMillis != null) {
                put(NativeSqliteHelper.COLUMN_EXPIRY, System.currentTimeMillis() + ttlMillis)
            } else {
                putNull(NativeSqliteHelper.COLUMN_EXPIRY)
            }
        }
        helper.writableDatabase.insertWithOnConflict(
            NativeSqliteHelper.TABLE,
            null,
            values,
            SQLiteDatabase.CONFLICT_REPLACE,
        )
        Unit
    }

    override fun getBytes(key: String): ByteArray? = uniffiCallbackGuard(ErrorType.CACHE) {
        helper.readableDatabase.query(
            NativeSqliteHelper.TABLE,
            arrayOf(NativeSqliteHelper.COLUMN_VALUE, NativeSqliteHelper.COLUMN_EXPIRY),
            "${NativeSqliteHelper.COLUMN_KEY} = ?",
            arrayOf(key),
            null,
            null,
            null,
        ).use { cursor ->
            if (!cursor.moveToFirst()) return@uniffiCallbackGuard null
            val expiryIndex = cursor.getColumnIndexOrThrow(NativeSqliteHelper.COLUMN_EXPIRY)
            if (!cursor.isNull(expiryIndex)) {
                val expiry = cursor.getLong(expiryIndex)
                if (System.currentTimeMillis() > expiry) {
                    remove(key)
                    return@uniffiCallbackGuard null
                }
            }
            cursor.getBlob(cursor.getColumnIndexOrThrow(NativeSqliteHelper.COLUMN_VALUE))
        }
    }

    override fun remove(key: String) = uniffiCallbackGuard(ErrorType.CACHE) {
        helper.writableDatabase.delete(
            NativeSqliteHelper.TABLE,
            "${NativeSqliteHelper.COLUMN_KEY} = ?",
            arrayOf(key),
        )
        Unit
    }

    override fun clear() = uniffiCallbackGuard(ErrorType.CACHE) {
        helper.writableDatabase.delete(NativeSqliteHelper.TABLE, null, null)
        Unit
    }

    override fun has(key: String): Boolean = getBytes(key) != null

    override fun dispose() = uniffiCallbackGuard(ErrorType.CACHE) {
        helper.close()
    }

}

File: platforms/ios/runtime/Sources/NativeblocksRuntime/lib/NativeSqliteStore.swift
import Foundation
import NativeblocksRuntimeFFI
import SQLite3

private let SQLITE_TRANSIENT = unsafeBitCast(-1, to: sqlite3_destructor_type.self)

internal final class NativeSqliteHelper {

    private static let table = "nativeblocks_cache"
    private static let columnKey = "cache_key"
    private static let columnValue = "cache_value"
    private static let columnExpiry = "cache_expiry"

    private let lock = NSLock()
    private var handle: OpaquePointer?

    init(databaseName: String) throws {
        let directory = try FileManager.default.url(
            for: .applicationSupportDirectory,
            in: .userDomainMask,
            appropriateFor: nil,
            create: true
        )
        try FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true)
        let path = directory.appendingPathComponent("\(databaseName).sqlite").path

        var db: OpaquePointer?
        guard sqlite3_open(path, &db) == SQLITE_OK, let opened = db else {
            let reason = db.map { String(cString: sqlite3_errmsg($0)) } ?? "unknown error"
            sqlite3_close(db)
            throw NbError.Failure(reason: "Cannot open \(path): \(reason)", errorType: .cache, errorCode: nil)
        }
        self.handle = opened

        try execute(
            """
            CREATE TABLE IF NOT EXISTS \(Self.table) (
                \(Self.columnKey) TEXT PRIMARY KEY,
                \(Self.columnValue) BLOB NOT NULL,
                \(Self.columnExpiry) INTEGER
            )
            """
        )
    }


    func save(key: String, value: Data, ttlMillis: Int64?) throws {
        lock.lock()
        defer { lock.unlock() }

        let sql = """
            INSERT OR REPLACE INTO \(Self.table)
            (\(Self.columnKey), \(Self.columnValue), \(Self.columnExpiry)) VALUES (?, ?, ?)
            """
        let statement = try prepare(sql)
        defer { sqlite3_finalize(statement) }

        sqlite3_bind_text(statement, 1, key, -1, SQLITE_TRANSIENT)
        _ = value.withUnsafeBytes { buffer in
            sqlite3_bind_blob(statement, 2, buffer.baseAddress, Int32(buffer.count), SQLITE_TRANSIENT)
        }
        if let ttlMillis {
            sqlite3_bind_int64(statement, 3, Self.nowMillis() + ttlMillis)
        } else {
            sqlite3_bind_null(statement, 3)
        }

        guard sqlite3_step(statement) == SQLITE_DONE else { throw failure() }
    }

    func load(key: String) throws -> Data? {
        lock.lock()

        let sql = """
            SELECT \(Self.columnValue), \(Self.columnExpiry) FROM \(Self.table)
            WHERE \(Self.columnKey) = ?
            """
        let statement: OpaquePointer?
        do {
            statement = try prepare(sql)
        } catch {
            lock.unlock()
            throw error
        }
        sqlite3_bind_text(statement, 1, key, -1, SQLITE_TRANSIENT)

        guard sqlite3_step(statement) == SQLITE_ROW else {
            sqlite3_finalize(statement)
            lock.unlock()
            return nil
        }

        if sqlite3_column_type(statement, 1) != SQLITE_NULL {
            let expiry = sqlite3_column_int64(statement, 1)
            if Self.nowMillis() > expiry {
                sqlite3_finalize(statement)
                lock.unlock()
                // Evicting takes the lock again, hence the unlock above.
                try remove(key: key)
                return nil
            }
        }

        defer {
            sqlite3_finalize(statement)
            lock.unlock()
        }
        guard let bytes = sqlite3_column_blob(statement, 0) else { return Data() }
        return Data(bytes: bytes, count: Int(sqlite3_column_bytes(statement, 0)))
    }

    func remove(key: String) throws {
        lock.lock()
        defer { lock.unlock() }

        let statement = try prepare("DELETE FROM \(Self.table) WHERE \(Self.columnKey) = ?")
        defer { sqlite3_finalize(statement) }

        sqlite3_bind_text(statement, 1, key, -1, SQLITE_TRANSIENT)
        guard sqlite3_step(statement) == SQLITE_DONE else { throw failure() }
    }

    func clear() throws {
        lock.lock()
        defer { lock.unlock() }
        try execute("DELETE FROM \(Self.table)")
    }

    func close() {
        lock.lock()
        defer { lock.unlock() }
        guard let handle else { return }
        sqlite3_close(handle)
        self.handle = nil
    }

    deinit {
        if let handle { sqlite3_close(handle) }
    }

    private func prepare(_ sql: String) throws -> OpaquePointer? {
        guard let handle else {
            throw NbError.Failure(reason: "The cache database is closed", errorType: .cache, errorCode: nil)
        }
        var statement: OpaquePointer?
        guard sqlite3_prepare_v2(handle, sql, -1, &statement, nil) == SQLITE_OK else {
            sqlite3_finalize(statement)
            throw failure()
        }
        return statement
    }

    private func execute(_ sql: String) throws {
        let statement = try prepare(sql)
        defer { sqlite3_finalize(statement) }
        guard sqlite3_step(statement) == SQLITE_DONE else { throw failure() }
    }

    private func failure() -> NbError {
        let reason = handle.map { String(cString: sqlite3_errmsg($0)) } ?? "the cache database is closed"
        return NbError.Failure(reason: reason, errorType: .cache, errorCode: nil)
    }

    private static func nowMillis() -> Int64 {
        return Int64(Date().timeIntervalSince1970 * 1000)
    }
}

internal final class SqliteCacheProvider: CacheProvider, @unchecked Sendable {

    private let helper: NativeSqliteHelper

    init(helper: NativeSqliteHelper) {
        self.helper = helper
    }

    func saveBytes(key: String, value: Data, ttlMillis: Int64?) throws {
        try uniffiCallbackGuard(.cache) {
            try helper.save(key: key, value: value, ttlMillis: ttlMillis)
        }
    }

    func getBytes(key: String) throws -> Data? {
        try uniffiCallbackGuard(.cache) {
            try helper.load(key: key)
        }
    }

    func remove(key: String) throws {
        try uniffiCallbackGuard(.cache) {
            try helper.remove(key: key)
        }
    }

    func clear() throws {
        try uniffiCallbackGuard(.cache) {
            try helper.clear()
        }
    }

    func has(key: String) throws -> Bool {
        return try getBytes(key: key) != nil
    }

    func dispose() throws {
        try uniffiCallbackGuard(.cache) {
            helper.close()
        }
    }
}
